design/disaster-recovery-snapshots.md

```diff
@@ -0,0 +1,128 @@
+# Optionally enable disaster recovery snapshots
+
+Currently KOTS supports application level snapshots, which can be used to restore a previously deployed application version along with its data.
+However in case of a disaster, only application can be restored to a new cluster and Admin Console can never be used to manage it again.
+This document proposes a way to implement disaster recovery snapshots using Velero.
+
+## Goals
+
+1. Use Velero to create Admin Console snapshot.
+1. Make CLI based disaster recovery possible.
+1. Propose a way to enable UI based recovery process.
+
+## Non Goals
+
+1. Improving application deployment process to avoid the `field is immutable` errors.
+1. Performing incremental backups of docker images.
+
+## Background
+
+Velero allows a number of ways to include and exclude resources in backups.
+In current implememntation, backup is created with the `includedNamespaces` key set to include all namespaces where application is deployed.
+All Admin Console resources have label `velero.io/exclude-from-backup: "true"`, which is recognized by Velero to exclude these resources.
+
+
+Disaster recovery snapshots must include Admin Console data, and potentially all resources.
+Due to this requirement, `velero.io/exclude-from-backup` label can no longer be used.
+The `includedNamespaces` filter also cannot be used because Admin Console and application can be deployed into the same namespace, which will cause problems during the restore process.
+
+## High-Level Design
+
+Velero backup spec supports the `labelSelector` field.
+All resources, application and Admin Console, can be labeled for inclusion in snapshots.
+Similarly, restore spec supports the same `labelSelector` field to select specific resources to restore.
+Using different `labelSelector` values allows to restore application and Admin Console separately.
+
+## Detailed Design
+
+### Backup
+
+- All resources (Admin Console and application) will have an additional label:
+    - `kots.io/backup: velero`
+- All application resources will have two additional labels:
+    - `kots.io/backup: velero`
+    - `kots.io/app-slug: <slug>`
+- Backup will be created with `labelSelector` set as in the example below:
+
+        apiVersion: velero.io/v1
+        kind: Backup
+        spec:
+          labelSelector:
+            matchLabels:
+              kots.io/backup: velero
+
+### Restore
+
+- Admin Console restore will be created with `labelSelector` set as in the example below:
+
+        apiVersion: velero.io/v1
+        kind: Restore
+        spec:
+          labelSelector:
+            matchLabels:
+              kots.io/kotsadm: "true"
+
+- Application restore will be created with `labelSelector` set as in the example below:
+
+        apiVersion: velero.io/v1
+        kind: Restore
+        spec:
+          labelSelector:
+            matchLabels:
+              kots.io/app-slug: <slug>
```

Comment:
I like the ability to select all objects related to an app in other ways, not just for backup/restore. When we add the ability to delete an app, for example, we should be able to make use of the same label (assuming apps don't create new objects that aren't tracked).

Reply:
It could possibly be used to check if we're about to overwrite an existing object with an application-specific version. Say, if two apps both try to install nginx in the same namespace. Not sure what we'd expect the user to do with that information, but it could at least be a warning.